### PR TITLE
OpenAI: Avoid two streams to overlap when doing multiple requests

### LINKED ISSUE #75179: DashGPT: When user triggers multiple generation calls, the streams conflict
When a user triggers simultaneous calls to generate content via GPT the streams can conflict / override each other. On initial testing this appears to only impact a single user and not several users in the same instance (thankfully)

https://github.com/grafana/grafana/assets/22381771/0c210b01-c59a-4726-a7e7-7b849e676aa1

### PR DESCRIPTION
**Pull request by SandersAaronD**
**What is this feature?**

Fixes frontend issue for DashGPT to allow multiple calls to OpenAI to run without interfering with each other.

**Why do we need this feature?**

Right now if a user runs multiple requests to OpenAI the output from the first to return will end up in the field for the second one.

**Who is this feature for?**

Anyone using DashGPT.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/75179

(It was fixed in @grafana/experimental, and then we got a new release of @grafana/experimental, and for some reason that was broken, so I moved the code over here to try to make sure we can actually deploy it. It's just some utils, so should be fine.)

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by leeoniya** [comment]
it's best not to [cross streams](https://www.urbandictionary.com/define.php?term=Crossing%20Streams) 🤣

**Review by ivanortegaalba — APPROVED** [review]
@SandersAaronD looks good so far!

Moving here `openai` utils will help to develop quicker. Although, if this keeps moving, or to be able to consume this from a plugin, it should live and being exposed as package (but we will get into the issue of having to wait until Grafana is released to get changes for the plugin, so probably the best option would be a totally independent package)

File: public/app/features/dashboard/components/GenAI/GenAIDashDescriptionButton.tsx [hunk 1]
```diff
@@ -26,23 +30,32 @@ export const GenAIDashDescriptionButton = ({ onGenerate, dashboard }: GenAIDashD
...
+      ${panel.description ? `- Description: ${panel.description}` : ''}
+      `
+  )
+  let panelString: string;
+  if (panels.length <= 10) {
+    panelString = `The the panels in the dashboard are:\n${panels.join('\n')}`;
+  } else {
+    panelString = `There are ${panels.length} panels.\n
+    Due to space constraints, only the information for the first ten are how is presented.\n
+    The the panels in the dashboard are:\n${panels.slice(10).join('\n')}`
```
[the remark this answers is not in the document]
Comment (SandersAaronD) [reply public/app/features/dashboard/components/GenAI/GenAIDashDescriptionButton.tsx]: Sounds good, I'll try to get that implemented in just a bit and then we can merge this in. Once it's in (with the OAI utils) we can work on getting error propagation to work nicely.

File: public/app/features/dashboard/components/GenAI/llms/openai.ts [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+export type ChatCompletionsDelta = ContentMessage | FunctionCallMessage | DoneMessage;
+
+/** A chunk included in a chat completion response. */
+export interface ChatCompletionsChunk {
+  /** The delta since the previous chunk. */
+  delta: ChatCompletionsDelta;
+}
+
+/** Return true if the message is a 'content' message. */
+export function isContentMessage(message: unknown): message is ContentMessage {
```
Comment (SandersAaronD) [inline public/app/features/dashboard/components/GenAI/llms/openai.ts:234]: I cannot figure out how on earth to make the linter accept these functions. Parallel change in experimental pending here: https://github.com/grafana/grafana-experimental/pull/87/files

File: public/app/features/dashboard/components/GenAI/llms/openai.ts [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+/** Return true if the message is a 'done' message. */
+export function isDoneMessage(message: unknown): message is DoneMessage {
+  return (message as DoneMessage).done !== undefined;
+}
+
+/** Return true if the response is an error response. */
+export function isErrorResponse(response: unknown): response is ChatCompletionsErrorResponse {
+  return (response as ChatCompletionsErrorResponse).error !== undefined;
+}
```
Comment (SandersAaronD) [inline public/app/features/dashboard/components/GenAI/llms/openai.ts:246]: The linter doesn't like these functions and I cannot figure out any variation on them that it considers acceptable

File: public/app/features/dashboard/components/GenAI/GenAIDashDescriptionButton.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (SandersAaronD) [reply public/app/features/dashboard/components/GenAI/GenAIDashDescriptionButton.tsx]: This is good on the current version of the PR.

File: public/app/features/dashboard/components/GenAI/llms/openai.ts (hunk 3 above)
Comment (SandersAaronD) [reply public/app/features/dashboard/components/GenAI/llms/openai.ts:246]: Similar PR here in experimental also not acceptable to the linter: https://github.com/grafana/grafana-experimental/pull/87

**Review by nmarrs — APPROVED** [review]
Functionality wise this is working great - not as excited about moving a bunch of llm code into core though 🤔 is this a long term solution or more or less temporary until we can figure out how to make it work with the code living in experimental?

**Comment by ivanortegaalba** [comment]
> Functionality wise this is working great - not as excited about moving a bunch of llm code into core though 🤔 is this a long term solution or more or less temporary until we can figure out how to make it work with the code living in experimental?

It is for the feedback gathering and active development. I'd move everything to experimental before GA (8th Oct)

File: public/app/features/dashboard/components/GenAI/llms/openai.ts (hunk 3 above)
Comment (ivanortegaalba) [reply public/app/features/dashboard/components/GenAI/llms/openai.ts:246]: Yep, I did some changes, but honestly, I don't know how to do better the type assertion here 😓

**Pull request merged**